scio-neo4j/src/main/scala/com/spotify/scio/neo4j/ops/Neo4jSCollectionReadOps.scala

```diff
@@ -0,0 +1,149 @@
+ * under the License.
+ */
+
+package com.spotify.scio.neo4j.ops
+
+import com.spotify.scio.coders.{Coder, CoderMaterializer}
+import com.spotify.scio.neo4j.Neo4jOptions
+import com.spotify.scio.neo4j.ops.Neo4jCommon.dataSourceConfiguration
+import com.spotify.scio.values.SCollection
+import magnolify.neo4j.ValueType
+import org.apache.beam.sdk.io.{neo4j => beam}
+import org.apache.beam.sdk.transforms.SerializableFunction
+
+import java.util.{Map => JMap}
+import scala.jdk.CollectionConverters._
+
+/**
+ * Operations for transforming an existing [[SCollection]] via a Neo4j query (read)
+ * @param self
+ *   existing [[SCollection]] (prior to Neo4j read)
+ * @tparam X
+ *   type parameter for existing [[SCollection]]
+ */
+class Neo4jSCollectionReadOps[X](self: SCollection[X]) {
+
+  /**
+   * Execute parallel instances of the provided Cypher query to the specified Neo4j database; one
+   * instance of the query will be executed for each element in this [[SCollection]]. Results from
+   * each query invocation will be added to the resulting [[SCollection]] as if by a `flatMap`
+   * transformation (where the Neo4j-query-execution returns an `Iterable`).
+   *
+   * This operation parameterizes each query invocation by applying a supplied function to each
+   * [[SCollection]] element before executing the Cypher query. The function must produce a [[Map]]
+   * of [[String]] to [[AnyRef]], where the keys correspond to named parameters in the provided
+   * Cypher query and the corresponding values correspond to the intended value of that parameter
+   * for a given query invocation. Named parameters must consist of letters and numbers, prepended
+   * with a `$`, as described in the Neo4j
+   * [[https://neo4j.com/docs/cypher-manual/current/syntax/parameters Cypher Manual (Syntax / Parameters)]].
+   *
+   * @see
+   *   ''Reading from Neo4j'' in the
+   *   [[https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/neo4j/Neo4jIO.html Beam `Neo4jIO` documentation]]
+   * @param neo4jConf
+   *   [[Neo4jOptions]] indicating the Neo4j instance on which to run the query
+   * @param cypher
+   *   [[String]] parameterized Cypher query
+   * @param elementToParamFunction
+   *   function (`X => Map[String, AnyRef]`) which converts an element of the [[SCollection]] to a
+   *   [[Map]] of parameter values for the given Cypher query
+   * @param neo4jType
+   *   (implicit) [[ValueType]] converting Neo4j results to the expected [[SCollection]] output type
+   * @param coder
+   *   (implicit) [[Coder]] for serialization of the result [[SCollection]] type
+   * @tparam Y
+   *   result [[SCollection]] type
+   * @return
+   *   [[SCollection]] containing the union of query results from a parameterized query invocation
+   *   for each original [[SCollection]] element
+   */
+  def neo4jCypherWithParams[Y](
+    neo4jConf: Neo4jOptions,
+    cypher: String,
+    elementToParamFunction: X => Map[String, AnyRef]
```

Comment:
It would look nicer if the function is curried.
In the write API, we actually user the `ValueType` to convert `T` to the params.
I think we can also have such API